Make location path queries more powerful

esoergel opened the pull request:
##### SUMMARY
This is something I've had in my to-do list for a long while, and I figured the wider adoption of the feature brought about by https://github.com/dimagi/commcare-hq/pull/26646 warrants the hour or so of work it took.

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### PRODUCT DESCRIPTION
This reworks the filtering logic when searching for a location.  It differs from the current implementation in that:
 - you can use more than one slash
- you don't need to use quotes to use slashes
- you can partial match ancestors (eg, "mass/mid/cam" to get "Massachusetts/Middlesex/Cambridge")
- quotes give you exact name matches, but still return descendant matches

This last decision may be controversial, and I'm open to reconsidering (say, if the last node is quoted, don't show descendants of it).  I don't like how this changes the logic, though, and think the point would be moot with proper ordering, which would show the first exact match at the top.

If accepted as-is, we should probably update https://confluence.dimagi.com/display/commcarepublic/Exact+Search+for+Locations

esoergel committed:
Make location path queries more powerful

mkangia commented on corehq/apps/locations/models.py:
```diff
@@ -310,15 +310,21 @@ def get_from_user_input(self, domain, user_input):
+        """
+        query = self
+        for part in filter(None, user_input.split('/')):
+            if part.startswith('"') and part.endswith('"'):
```
what happens when user is entering the name and has type the partial name with the starting quotes only?
for ex: "middle

we would search for `middle` as the default behaviour until the user puts the closing quotes?

kaapstorm replied:
@mkangia, yes exactly. `filter_by_user_input()` calls `_user_input_filter()`, so it preserves current behaviour until the final `"` is typed.

mkangia replied:
I think it used to do exact search even if you just have the first quote.

esoergel replied:
That's correct.  I prefer this behavior for two reasons - one, it's what google does, so arguably more familiar.  More importantly, this is a typeahead filter, so if the user has input `"Mass`, it's likely they just haven't finished the word yet, and it makes more sense to show `Massachusetts` than no results, unless they've put the closing quote, indicating that they _do_ want an exact match on `"Mass"`

esoergel replied:
(you can see this in test cases here: https://github.com/dimagi/commcare-hq/pull/26671/files#diff-4a69b47c01ede698484ccc328abbc5c4R192-R194)

mkangia replied:
Sounds good to me, just wanted to ensure that my understanding is correct that behaviour has now changed.
I just feel that when a user has decided to do an exact match and is typing the location name for exact search, anything the HQ shows in that time period is not useful for the user i.e if i am searching for Massachusetts, a person like me would type `"ma` and then will type slowly trying to remember what the spelling is, in that interim whatever result HQ is showing me, doing all the effort to find matches and descendants is not useful for me because the reason i am doing an exact search is because the default search isn't working for my search. It's just unnecessary search and display that the user is ignoring.
This probably will be reduced since you have changed to just locations by name and not their descendants included.

orangejenny replied:
> will type slowly trying to remember what the spelling is

This sounds similar to what Ethan's saying - if you don't remember the spelling, and we're showing similar results, you might notice your option in the list and then not need to spell it out. And if that's not the case and you do ignore those interim results, that's fine.

kaapstorm commented on corehq/apps/locations/tests/test_location_queries.py:
```diff
@@ -171,3 +155,47 @@ def test_filter_path_by_user_input(self):
+                ('Brookline', []),
+            ])
+        ]),
+        ('Copycat State', [
```
[Pack the Union](https://harvardlawreview.org/2020/01/pack-the-union-a-proposal-to-admit-new-states-for-the-purpose-of-amending-the-constitution-to-ensure-equal-representation/) :us: :eagle: :statue_of_liberty:

kaapstorm reviewed (approved):
Nice

esoergel commented:
I had another thought about this behavior randomly last night.  @mkangia I think you'd like it in particular.

In the old implementation of `filter_path_by_user_input`, typing "Mass" yields also children of Massachusetts.  This PR doesn't change that.  I like that because you filter what you see - the path to a location.  My understanding is this becomes difficult in practice when you do want to only search by the name.  The current EMWF implementation uses this special syntax to change the rules of searching, though IMO it's not especially intuitive.

This PR _does_ incorporate semantic slashes from the EMWF filter.  One possibility would be to make the last node of your search match only names, and allow for you to put a trailing slash to indicate you'd like to see children.  That is, rather than the current behavior:
> `suff` -> `Massachusetts/Suffolk, Massachusetts/Suffolk/Boston, ...`

Do 
> `suff` -> `Massachusetts/Suffolk`
> `suff/` -> `Massachusetts/Suffolk/Boston, Massachusetts/Suffolk/Brookline, ...`

That way quotes would still control only exactness, and slashes would let you control descendant queries.

esoergel commented:
As implemented in this PR:
![image](https://user-images.githubusercontent.com/2367539/74786162-01ecb800-527a-11ea-84e3-4efd56c88d2e.png)

With my proposed further change:
![image](https://user-images.githubusercontent.com/2367539/74786233-2fd1fc80-527a-11ea-8685-21928275db0d.png)

And with a slash:
![image](https://user-images.githubusercontent.com/2367539/74786276-4b3d0780-527a-11ea-8406-18a5b3a4472e.png)


The main thing here is that if the user provides no special syntax, it only matches on location name (not ancestry).

mkangia commented:
Thanks for attaching those images @esoergel , graphical examples are always appreciated.

kaapstorm commented:
That sounds like a great idea, @esoergel

esoergel committed:
Further tweak search algorithm

As described in https://github.com/dimagi/commcare-hq/pull/26671#issuecomment-587954514

esoergel committed:
Update help text to describe new behavior

esoergel committed:
Update test for new dependant logic

esoergel committed:
Remove no-longer-used filter_by_user_input

esoergel committed:
Drop "path" from method name

kaapstorm commented on corehq/apps/locations/tests/test_location_queries.py:
```diff
@@ -181,21 +181,23 @@ class TestFilterPath(LocationHierarchyTestCase):
         for querystring, expected in [
-            ('Suffolk', ['Suffolk', 'Boston', 'Brookline']),
-            ('Suff', ['Suffolk', 'Boston', 'Brookline']),
+            ('Suff', ['Suffolk']),
```
Nit: Might be nice to show what an empty `querystring` does:
```python
            ('', ['Massachusetts', 'Middlesex', 'Cambridge', 'Somerville', 'Evil Somerville',
                  'Suffolk', 'Boston', 'Brookline', 'Copycat State', 'Evil Middlesex', 'Cambridge',
                  'Somerville'])
```

esoergel replied:
https://github.com/dimagi/commcare-hq/pull/26699
